[WIP] Sqlite Database engine for asset cache

rajarshimaitra opened the pull request:
This closes [Issue#58](https://github.com/LNP-BP/rgb-node/issues/58). This is WIP and intended for initial review and suggestions for possible modifications.

**Note:** 
Two test cases are added to test the working of an asset db. These tests are ignored by default as they will break travis build.
To test the DB do the followings: 

- Install diesel-cli by running `cargo install diesel-cli`
- Setup an environment variable pointing at the rgb datadir `export DATABASE_URL=~/.rgb`
- Manually remove `#[ignore]` flag from tests in `src/contract/fungible/cache/sql.rs`.
- run the testing script in `test/test_db.sh`.

This should run the tests, and if all goes well they should pass.

**Rough Edges**

- `Chain` variants are not fully covered. The database only supports `MainNet` and `Testnet3` for now as they can be encoded via text. The rest of the variants need more expressive table types or probably new tables to handle them. Suggestions are welcome. At least if we can do `Regtest` it will be usable for experimenting with real rgb-node.

**Possible Improvements**  
Once this PR gets finalized it would better to integrate Travis to run these tests. They have to be handled separately from regular tests as it involves running `diesel-cli` and setting up env variable.

dr-orlovsky reviewed (changes requested):
The overall logic looks good. I have few things that should be addressed globally (I am not commenting each case throughout the code and giving generic guidelines here):

1. `unwraps` are prohibited; and there is too many of them. If you quite sure that the function will never fail, use `expect` with the explanation why it would never fail; otherwise the error should be propagated upstream and error types must be changed/redesigned to handle that propagation.
2. `Chain` must be converted into string using it's native `FromStr/Display` support, no need to invent any new conversions or write any functions. Just `from_str()` and `to_string()`, and that's all.
3. For type conversion use `From` or `TryFrom` (for failiable conversions) traits, not custom functions. These must take the argument by value, not reference (also rust API guideline), and this will prevent many `clone()`s inside the code
4. Another rust API recommendation is to avoid repetition of the mod (i.e. file) name in the type name. For instance, it should be `sql::Asset` rather than `sql::SqlAsset` etc. Also error variants should not end with `Error`, since it is in type name aleary (I am having this issue with my own errors, so feel free to change mine parts as well). 
5. No need to add a new variant to `DataFormat`; this type is designed for formatting input and output for `cli` (command-line tool) and has no relation to `fungible`. The extensions for database should be hardcoded (there is only a one option for it) - pls see comments in the code.
6. You do not need to write custom convertors for error types; use `#[from(<type>)]` attribute and that's all. You can have multiple `#[from]`s for a single enum variant.
7. I am very cautious of saving the whole database each time; this will make support for Postgress really hard in the future and will trigger all stored procedures for all assets each time in case of exchanges. So we need to add some field `dirty` to the data structs, which should not be mapped and serialized into DB or file cache, but should be used to save only those items which are marked with this field.

Have not checked the tests yet; will do that later.

I do recommend to use CLion with rust plugin for simple renamings, since it allows to rename type, variant or fn in one place ans automatically applies that everywhere. If you do not have it, just let me know and leave all renamings to me, I will do them really quickly saving lot of your time.

rajarshimaitra commented on src/contracts/fungible/data/asset.rs:
```diff
@@ -134,6 +144,39 @@ pub struct Asset {
+        table_value: &SqlAsset,
+        connection: &SqliteConnection,
+    ) -> Result<Self, SqlCacheError> {
+        let (known_inflation, unknown_inflation) = read_inflation(table_value, connection).unwrap();
```
We should not, fixed.

rajarshimaitra commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -10,3 +10,687 @@
+
+    #[inline]
+    pub fn assets_filename(&self) -> PathBuf {
+        self.assets_dir()
+            .join("assets")
+            .with_extension(self.data_format.extension())
```
Done

rajarshimaitra commented:
>unwraps are prohibited; and there is too many of them. If you quite sure that the function will never fail, use expect with the explanation why it would never fail; otherwise the error should be propagated upstream and error types must be changed/redesigned to handle that propagation.

Yes i intended to remove all the unwraps. Few I missed in data module. Those are fixed now. Few new error types are added in cache error. Now the only unwraps remaining are in the test module, which I guess should be fine.

 > Chain must be converted into string using it's native FromStr/Display support, no need to invent any new conversions or write any functions. Just from_str() and to_string(), and that's all.

Noted and modified. 

>For type conversion use From or TryFrom (for failiable conversions) traits, not custom functions. These must take the argument by value, not reference (also rust API guideline), and this will prevent many clone()s inside the code

Initially I tried with From implementation. But in most of the cases, the conversion between Asset data and Sql Asset table entry are not one to one. They need inputs of other table entries to create the right entry. This is violating the From trait function signature. So the next best thing I could do is to have `impl` defined on both the structures. But this also doesn't work for `AllocationUtxo` and `Allocation` table entries, as they need to be read from and created simultaneously as the tables are associated. So I kept `create_allocation_from_asset()` and `read_allocations()` as stand-alone functions. Moreover, these are not part of the `SqlCache` API, and kind of like helper functions that the `impl` conversion logic uses.

>Another rust API recommendation is to avoid repetition of the mod (i.e. file) name in the type name. For instance, it should be sql::Asset rather than sql::SqlAsset etc. Also error variants should not end with Error, since it is in type name aleary (I am having this issue with my own errors, so feel free to change mine parts as well).

Tried this one too, but the `data:asset` and `sql::asset` are getting ambiguous when they are both used together, which is a lot. And as these are not part of library structures and module users will never use them directly, they are internal to `SqlCache`, I went with `SqlAsset`. We can have any other name to define them or just use explicity declaration like `sql::asset` and 'data::asset` everywhere.

>No need to add a new variant to DataFormat; this type is designed for formatting input and output for cli (command-line tool) and has no relation to fungible. The extensions for database should be hardcoded (there is only a one option for it) - pls see comments in the code.

This is fixed.

>You do not need to write custom convertors for error types; use #[from(<type>)] attribute and that's all. You can have multiple #[from]s for a single enum variant.

Didn't get what you meant here? If you are talking about removing the `#[from]` above the error variants, that's giving some "unimplemented From Trait" kinda errors.

 >I am very cautious of saving the whole database each time; this will make support for Postgress really hard in the future and will trigger all stored procedures for all assets each time in case of exchanges. So we need to add some field dirty to the data structs, which should not be mapped and serialized into DB or file cache, but should be used to save only those items which are marked with this field.

I also don't feel this is the best thing to. But I did it with the first draft because it was easy and it worked for small datasets. So I think this even might work for wallet dbs. But for high load works something better needs to be done. 

Regarding Postgres, I don't think the same cache can be used. Though probably the same table schema and all the table related structures can be used (there might be some incompatibility there also, not sure). We would need to have a new cache anyway and that can have a better update handling mechanism.

Rest of the stuff I have fixed and updated and will push shortly.

dr-orlovsky commented on src/contracts/fungible/data/asset.rs:
```diff
@@ -134,6 +144,39 @@ pub struct Asset {
+    /// and recreates the full Asset structure. This should be used while reading Asset data from
+    /// database table.
+    pub fn from_sql_asset(
+        table_value: &SqlAsset,
```
You are correct, this will not reduce number of clones inside the function, so it's only about following convention for `from_*` methods. While the only thing that requires clone are the string values for metafields and the rest is either `Copy` or passed by reference to underlying functions, the strings are anyway the only place where we have clones so far. So let's leave it as is.

dr-orlovsky commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -10,3 +10,683 @@
+    NotFound,
+}
+
+impl From<SqlCacheError> for ServiceErrorDomain {
+    fn from(_: SqlCacheError) -> Self {
+        ServiceErrorDomain::Cache
+    }
+}
+
+impl From<SqlCacheError> for BootstrapError {
+    fn from(_: SqlCacheError) -> Self {
+        BootstrapError::CacheError
+    }
+}
```
This can be removed and replaced with two `#[from]` directives in `ServiceErrorDomain` declaration

rajarshimaitra replied:
Good point. Should it also work for `FileCacheError`? And how should i write conversion derivation from multiple error types?

dr-orlovsky replied:
Conversion for multiple types works with multiple and separate `#[from(...)]` attributes.

rajarshimaitra replied:
And should the same thing could be done for `FileCacheError`?

dr-orlovsky replied:
Yes

rajarshimaitra replied:
Done. Added the same for filecache error too.

dr-orlovsky reviewed (changes requested):
Seems like we are sorted. Haven't read tests yet, will do a bit later.

1. Unwraps in tests are fine, yes;
2. Agree on `from_*`s, missed that point myself
3. Naming with `Sql..`: I see your point. For me it's fine to use `data::asset` and `sql::asset` (like `secp256k1::PublicKey` and `bitcoin::PublicKey`), but I agree that since this is just for internal use in the mod it will be easier to have `Asset` and `SqlAsset`.
4. Regarding `#[from]` made a comment in the text. This is the final point that remains to change
5. We will do better DB update later, agree. The question is why we can't use the same code for Postgres?

rajarshimaitra commented on src/contracts/fungible/cache/sql.rs:
```diff
@@ -72,20 +77,17 @@ impl From<SqlCacheError> for BootstrapError {
-        self.assets_dir()
-            .join("assets")
-            .with_extension(self.data_format.extension())
+        self.assets_dir().join("assets").with_extension("db")
```
I am little confused here. Whats should the `datadir` look like? Is it like `../datadir/assets.db` or `../datadir/assets/assets.db? I think I did the second one now.

dr-orlovsky replied:
The second. But what you did is `../datadir/assets/assets/assets.db`

dr-orlovsky replied:
Oh right, you are right. Assets is the file name here, not directory

rajarshimaitra replied:
yes.

rajarshimaitra commented:
>We will do better DB update later, agree. The question is why we can't use the same code for Postgres?

For one thing the cache will need a new connection handler for postgress (`PglConnection`). This connection type is to be passed among all the helper functions that reads the table to create datastructures. Though I think this can be handled by some Enum and matching logics. The worst incompatibility is if there's any in the table datatype. But I hope as we are only using standard SQL datatypes, they should work in PGL too. Only the connection logic has to be handled properly. 

Fixed the `#[from()]` conversion between error types, and I think this is now ready for final review.

dr-orlovsky commented:
Great work overall, merging now
